proxy/rpc/np_gas_price.py

```diff
@@ -140,3 +175,169 @@ async def get_neon_gas_price(
+        avg_block_cu_price /= block_cnt * (block_cnt + 1) / 2
+
+        # Convert it into ethereum world by multiplying by suggested_gas_price
+        # N.B. prices in the block are stored in microlamports, so conversion to lamports takes place.
+        _, token_gas_price = await self._get_token_gas_price(ctx)
+        return int(token_gas_price.suggested_gas_price * avg_block_cu_price / 1_000_000)
+
+    @NeonProxyApi.method(name="eth_feeHistory")
+    async def get_fee_history(
+        self,
+        ctx: HttpRequestCtx,
+        block_cnt: HexUIntField,
+        block_tag: RpcBlockRequest,
+        priority_fee_percentile_list: list[int] | None,
+    ) -> _RpcFeeHistoryResp | None:
+        if not priority_fee_percentile_list:
+            # In case the User was lazy to request any percentiles, let's return "median"
+            # so the response makes sense.
+            priority_fee_percentile_list = [50]
+
+        # Validate input parameters, throw EthError if those are incorrect.
+        self._validate_fee_history_inputs(block_cnt, priority_fee_percentile_list)
+
+        # Fetch the current gas price - it's needed to convert price to gas tokens
+        # and to return base_fee_per_gas for the upcoming block.
+        _, token_gas_price = await self._get_token_gas_price(ctx)
+        current_gas_price: int = token_gas_price.suggested_gas_price
+
+        # Fetch the latest block slot.
+        block = await self.get_block_by_tag(block_tag)
+        # In case block is latest and it's empty, let's use this max value
+        # instead of making an extra get_latest_block DB query.
+        latest_slot = _MAX_LATEST_BLOCK
+        if not block.is_empty:
+            latest_slot = block.slot
+
+        # Let's clamp num_blocks to not create excessive load - Infura and Alchemy currently do the same.
+        block_cnt = min(block_cnt, 1024)
+        # Fetching the data stage.
+        fee_gas_data_list: list[BlockFeeGasData] = await self._db.get_historical_base_fees(
+            self._get_chain_id(ctx), block_cnt, latest_slot
+        )
+        cu_price_data_list: list[PriorityFeePercentiles] = await self._db.get_historical_priority_fees(
+            block_cnt, latest_slot
+        )
+
+        # Response objects.
+        base_fee_list: list[int] = []
+        gas_used_ratio_list: list[float] = []
+        reward_list: list[list[int]] = []
+
+        # Iterate through the fee_gas_data_list and cu_price_data_list using two pointers.
+        # Both lists are sorted in the descending by block_slot order.
+        # Also, fee_gas_data_list may miss entries for some blocks. In such a case, let's fill the output
+        # with the previous base_fee_per_gas we encounter (or with 0 until we see the first one).
+        fee_data_it: int = len(fee_gas_data_list) - 1
+        cu_data_it: int = len(cu_price_data_list) - 1
+        last_base_fee_per_gas: int = 0
+        while cu_data_it >= 0:
```

Comment:
@afalaleev The logic here is the following:

we get transactions of the new type together with the `block_slot` (gas_fee_data).
we get priority fee percentiles of the solana (from all transactions) together with the `block_slot` - (cu_price_data).

Now, we "merge" those entries by the `block_slot`, so we have the consistent response (base_fee and reward correspond to each other, as they are from the same solana-neon blocks).

The problem I see is the following: most of the time, all entries from `gas_fee_data` will be with the `block_slot` smaller than the ones from `cu_price_data` (because of lots of empty neon blocks). 
As a result, we return zeroes as the base_fee in the response.

As an idea: maybe it makes sense to not merge by the `block_slot` and treat them separately (construct base_fee in the response by relying **solely** on the `gas_fee_data`; construct `reward` in the response by relying **solely** on the `cu_price_data`? This way, we lose consistency of the response, but in contrast the response makes more sense (without zeroes, with **actual** base_fee from neon blocks, and with rewards from solana blockchain). WDYT?

Reply:
Implemented as discussed f2f: expose recent token gas price from the mempool and use it in the eth_feehistory in case User asks for the recent block data.
In case User requests historical data - take the data from the DB from the transactions as it was already implemented.